rosbag2_tests/test/rosbag2_tests/test_rosbag2_cpp_get_service_info.cpp

```diff
@@ -0,0 +1,338 @@
+    }
+    return ready;
+  }
+
+  // relative path to the root of the bag file.
+  std::filesystem::path root_bag_path_;
+  std::future<void> node_spinner_future_;
+  std::atomic_bool exit_from_node_spinner_{false};
+};
+
+TEST_P(Rosbag2CPPGetServiceInfoTest, get_service_info_for_bag_with_topics_only) {
+  const std::string storage_id = GetParam();
+  const auto bag_path_str = get_bag_path_str();
+  {
+    // Create an empty bag with default storage
+    rosbag2_cpp::Writer writer;
+    rosbag2_storage::StorageOptions storage_options;
+    storage_options.storage_id = storage_id;
+    storage_options.uri = bag_path_str;
+    writer.open(storage_options);
+    test_msgs::msg::BasicTypes msg;
+    writer.write(msg, "test_topic", rclcpp::Time{});
+  }
+
+  rosbag2_cpp::Info info;
+  std::vector<std::shared_ptr<rosbag2_cpp::rosbag2_service_info_t>> ret_service_infos;
+  const std::string recorded_bag_uri = bag_path_str + "/" + get_bag_file_name();
+  ASSERT_NO_THROW(ret_service_infos = info.read_service_info(recorded_bag_uri, storage_id)) <<
+    recorded_bag_uri;
+
+  EXPECT_TRUE(ret_service_infos.empty());
+}
+
+TEST_P(Rosbag2CPPGetServiceInfoTest, get_service_info_for_bag_with_services_only) {
+  const std::string storage_id = GetParam();
+  const std::string bag_path_str = get_bag_path_str();
+
+  auto service_client_manager =
+    std::make_shared<rosbag2_test_common::ClientManager<test_msgs::srv::BasicTypes>>(
+    "test_service");
+
+  std::unique_ptr<rosbag2_cpp::writer_interfaces::BaseWriterInterface> writer_impl =
+    std::make_unique<SequentialWriterForTest>();
+  auto writer = std::make_unique<rosbag2_cpp::Writer>(std::move(writer_impl));
+
+  rosbag2_storage::StorageOptions storage_options;
+  storage_options.storage_id = storage_id;
+  storage_options.uri = bag_path_str;
+  rosbag2_transport::RecordOptions record_options =
+  {false, true, false, {}, {}, {"/rosout"}, {}, "cdr", 100ms};
+  auto recorder = std::make_shared<rosbag2_transport::Recorder>(
+    std::move(writer), storage_options, record_options);
+  recorder->record();
+
+  start_async_spin(recorder);
+  auto cleanup_process_handle = rcpputils::make_scope_exit(
+    [&]() {stop_spinning();});
+
+  ASSERT_TRUE(service_client_manager->wait_for_srvice_to_be_ready());
+  ASSERT_TRUE(wait_for_subscriptions(*recorder, {"/test_service/_service_event"}));
+
+  constexpr size_t num_service_requests = 3;
+  for (size_t i = 0; i < num_service_requests; i++) {
+    ASSERT_TRUE(service_client_manager->send_request());
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+  }
+
+  auto & writer_ref = recorder->get_writer_handle();
+  auto & recorder_writer =
+    dynamic_cast<SequentialWriterForTest &>(writer_ref.get_implementation_handle());
+
+  // By default, only client introspection is enabled.
+  // For one request, service event topic get 2 messages.
+  size_t expected_messages = num_service_requests * 2;
```

Comment:
i think this test work-around should be remove once we address https://github.com/ros2/rmw/issues/357? this is not exactly we want to maintain here, if i am not mistaken.

Reply:
No. For this test, only enable client introspection.  2 messages mean a request message and a response message. 

```c++
  explicit ClientManager(
    std::string service_name,
    size_t client_number = 1,
    bool service_event_contents = false,
    bool client_event_contents = true)
```